common/io.hpp

```diff
@@ -0,0 +1,58 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef MENDER_COMMON_IO_HPP
+#define MENDER_COMMON_IO_HPP
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+
+#include <system_error>
+#include <vector>
+
+namespace mender::common::io {
+
+using namespace std;
+
+using mender::common::expected::Expected;
+
+class Reader {
+public:
+	virtual ~Reader() {};
+
+	virtual Expected<int64_t, error_condition> Read(vector<byte> &dst) = 0;
```

Comment:
RFC about the error being returned: In `std` the only "official" error types are `error_condition` (platform independent) and `error_code` (platform dependent). The former comes with cross platform error codes so one can do things like `if (error.value() == no_such_file_or_directory)` and such, so it seems better for portability.

Then we have our own error type [here](https://github.com/mendersoftware/mender/blob/feature-c%2B%2B-client/common/error.hpp), which is template based, and where we can add a string as well, for context, which `error_condition` cannot do on its own (you can only get an error string for the code itself).

Now, there's an important property we have to satisfy, which is that `Reader` can not be a template, because `template` and `virtual` do not mix. So we have to decide on one, and only one, error type to be returned.

Here are some pros and cons I can think:

* Advantages when using ` error_condition`:
    * Does not need to allocate anything.
    * Uses only `std`. If we even assume that `std::expected` would be used in the future, then this would be an interface which needs no external types. Since this is an interface we have used *a lot* in Golang, maybe this makes it more reusable (not sure about that though)?
* Advantages when using `mender::common::error::Error<error_condition>`:
    * Can potentially describe the problem better by providing more context in the string.

I'm almost exactly 50/50 undecided on this one. What do you think?

Reply:
Gut feeling is leaning towards the 2nd option, but no deep insight or strong opinions here.

Reply:
I think I have a 51% leaning towards it as well, so if I see no further comments, I'll go for that.